File: doxa-api/src/main/kotlin/com/doxa/service/SyncService.kt
package com.doxa.service

import com.doxa.dto.*
import com.doxa.models.*
import com.doxa.repository.*
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

@Service
class SyncService(
    private val userRepository: UserRepository,
    private val postRepository: PostRepository,
    private val communityRepository: CommunityRepository,
    private val commentRepository: CommentRepository,
    private val userSavedPostRepository: UserSavedPostRepository,
    private val userCommunityRepository: UserCommunityRepository,
    private val userFollowRepository: UserFollowRepository
) {
    @Transactional
    fun pushSync(request: SyncPushRequest) {
        for (dto in request.users) {
            val user = userRepository.findById(dto.id).orElse(null)
            if (user != null) {
                user.name = dto.name
                user.email = dto.email
                user.password = dto.password
                user.handle = dto.handle
                user.bio = dto.bio
                user.avatarUrl = dto.avatarUrl
                user.bannerUrl = dto.bannerUrl
                user.followers = dto.followers
                user.following = dto.following
                user.isSynced = 1
                userRepository.save(user)
            } else {
                val newUser = User(
                    id = dto.id,
                    name = dto.name,
                    email = dto.email,
                    password = dto.password,
                    handle = dto.handle,
                    bio = dto.bio,
                    avatarUrl = dto.avatarUrl,
                    bannerUrl = dto.bannerUrl,
                    followers = dto.followers,
                    following = dto.following,
                    isSynced = 1
                )
                userRepository.save(newUser)
            }
        }

        for (dto in request.communities) {
            val community = communityRepository.findById(dto.id).orElse(null)
            if (community != null) {
                community.name = dto.name
                community.description = dto.description ?: ""
                community.members = dto.members
                community.creatorId = dto.creator_id
                community.bannerUrl = dto.banner_url
                community.isJoined = dto.is_joined
                community.isSynced = 1
                community.updatedAt = LocalDateTime.now()
                communityRepository.save(community)
            } else {
                val newCommunity = Community(
                    id = dto.id,
                    name = dto.name,
                    description = dto.description ?: "",
                    creatorId = dto.creator_id,
                    members = dto.members,
                    isJoined = dto.is_joined,
                    bannerUrl = dto.banner_url,
                    isSynced = 1,
                    createdAt = parseDateTime(dto.created_at),
                    updatedAt = parseDateTime(dto.updated_at)
                )
                communityRepository.save(newCommunity)
            }
        }

        for (dto in request.posts) {
            val post = postRepository.findById(dto.id).orElse(null)
            if (post != null) {
                post.userId = dto.user_id
                post.author = dto.author
                post.title = dto.title
                post.description = dto.description
                post.subject = dto.subject
                post.tag = dto.tag
                post.role = dto.role
                post.time = dto.time
                post.imageUrl = dto.image_url
                post.upvotes = dto.upvotes
                post.commentsCount = dto.comments_count
                post.isSaved = dto.is_saved
                post.isSynced = 1
                post.updatedAt = LocalDateTime.now()
                postRepository.save(post)
            } else {
                val newPost = Post(
                    id = dto.id,
                    userId = dto.user_id,
                    author = dto.author,
                    title = dto.title,
                    description = dto.description,
                    subject = dto.subject,
                    tag = dto.tag,
                    role = dto.role,
                    time = dto.time,
                    imageUrl = dto.image_url,
                    upvotes = dto.upvotes,
                    commentsCount = dto.comments_count,
                    isSaved = dto.is_saved,
                    isSynced = 1,
                    createdAt = parseDateTime(dto.created_at),
                    updatedAt = parseDateTime(dto.updated_at)
                )
                postRepository.save(newPost)
            }
        }

        for (dto in request.comments) {
            val comment = commentRepository.findById(dto.id).orElse(null)
            if (comment != null) {
                comment.text = dto.text
                comment.likes = dto.likes
                comment.isLiked = dto.isLiked
                comment.isSynced = 1
                comment.updatedAt = LocalDateTime.now()
                commentRepository.save(comment)
            } else {
                val newComment = Comment(
                    id = dto.id,
                    postId = dto.post_id,
                    userId = dto.user_id,
                    author = dto.author,
                    avatar = dto.avatar,
                    text = dto.text,
                    time = dto.time,
                    likes = dto.likes,
                    isLiked = dto.isLiked,
                    replies = dto.replies,
                    isSynced = 1,
                    createdAt = parseDateTime(dto.created_at),
                    updatedAt = parseDateTime(dto.updated_at)
                )
                commentRepository.save(newComment)
            }
        }

        for (rel in request.userSavedPosts) {
            val id = UserSavedPostId(rel.userId, rel.postId)
            if (!userSavedPostRepository.existsById(id)) {
                userSavedPostRepository.save(UserSavedPost(rel.userId, rel.postId))
            }
        }

        for (rel in request.userCommunities) {
            val id = UserCommunityId(rel.userId, rel.communityId)
            if (!userCommunityRepository.existsById(id)) {
                userCommunityRepository.save(UserCommunityEntity(rel.userId, rel.communityId))
            }
        }

        for (rel in request.userFollows) {
            val id = UserFollowId(rel.followerId, rel.followedId)
            if (!userFollowRepository.existsById(id)) {
                userFollowRepository.save(UserFollow(rel.followerId, rel.followedId))
            }
        }
    }

    @Transactional(readOnly = true)
    fun pullSync(): SyncPullResponse {
        val currentServerTime = LocalDateTime.now()

        val users = userRepository.findAll().map { user ->
            SyncUserDTO(
                id = user.id,
                name = user.name,
                email = user.email,
                password = user.password,
                handle = user.handle,
                bio = user.bio,
                avatarUrl = user.avatarUrl,
                bannerUrl = user.bannerUrl,
                followers = user.followers,
                following = user.following,
                is_synced = 1
            )
        }

        val posts = postRepository.findAll().map { post ->
            SyncPostDTO(
                id = post.id,
                user_id = post.userId,
                author = post.author,
                title = post.title,
                description = post.description,
                subject = post.subject,
                tag = post.tag,
                role = post.role,
                time = post.time,
                image_url = post.imageUrl,
                upvotes = post.upvotes,
                comments_count = post.commentsCount,
                is_saved = post.isSaved,
                created_at = post.createdAt.toString(),
                updated_at = post.updatedAt.toString(),
                is_synced = 1
            )
        }

        val communities = communityRepository.findAll().map { comm ->
            SyncCommunityDTO(
                id = comm.id,
                name = comm.name,
                members = comm.members,
                description = comm.description,
                is_joined = comm.isJoined,
                creator_id = comm.creatorId,
                banner_url = comm.bannerUrl,
                created_at = comm.createdAt.toString(),
                updated_at = comm.updatedAt.toString(),
                is_synced = 1
            )
        }

        val comments = commentRepository.findAll().map { comment ->
            SyncCommentDTO(
                id = comment.id,
                post_id = comment.postId,
                user_id = comment.userId,
                author = comment.author,
                avatar = comment.avatar,
                text = comment.text,
                time = comment.time,
                likes = comment.likes,
                isLiked = comment.isLiked,
                replies = comment.replies,
                is_synced = 1,
                created_at = comment.createdAt.toString(),
                updated_at = comment.updatedAt.toString()
            )
        }

        val userSavedPosts = userSavedPostRepository.findAll().map {
            UserPostRelation(it.userId, it.postId)
        }

        val userCommunities = userCommunityRepository.findAll().map {
            UserCommunityRelation(it.userId, it.communityId)
        }

        val userFollows = userFollowRepository.findAll().map {
            UserFollowRelation(it.followerId, it.followedId)
        }

        return SyncPullResponse(
            serverTimestamp = currentServerTime,
            users = users,
            posts = posts,
            communities = communities,
            comments = comments,
            userSavedPosts = userSavedPosts,
            userCommunities = userCommunities,
            userFollows = userFollows
        )
    }

    private fun parseDateTime(dateStr: String?): LocalDateTime {
        if (dateStr.isNullOrBlank()) return LocalDateTime.now()
        return try {
            LocalDateTime.parse(dateStr)
        } catch (e: Exception) {
            try {
                LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
            } catch (e2: Exception) {
                LocalDateTime.now()
            }
        }
    }
}

File: DOXA/services/syncService.ts
import { SQLiteDatabase } from 'expo-sqlite';
import { api } from './api';

export class SyncService {
    private db: SQLiteDatabase;

    constructor(database: SQLiteDatabase) {
        this.db = database;
    }

    async pushSync() {
        try {
            console.log("Iniciando Push Sync...");
            
            const unsyncedPosts = await this.db.getAllAsync<any>("SELECT * FROM posts WHERE is_synced = 0");
            const unsyncedCommunities = await this.db.getAllAsync<any>("SELECT * FROM communities WHERE is_synced = 0");
            const unsyncedComments = await this.db.getAllAsync<any>("SELECT * FROM comments WHERE is_synced = 0");
            const unsyncedUsers = await this.db.getAllAsync<any>("SELECT * FROM users WHERE is_synced = 0");

            console.log(`Found: ${unsyncedPosts.length} posts, ${unsyncedCommunities.length} communities, ${unsyncedComments.length} comments, ${unsyncedUsers.length} users to sync`);

            const userSavedPosts = await this.db.getAllAsync<any>("SELECT * FROM user_saved_posts");
            const userCommunities = await this.db.getAllAsync<any>("SELECT * FROM user_communities");
            const userFollows = await this.db.getAllAsync<any>("SELECT * FROM user_follows");

            const pushPayload = {
                posts: unsyncedPosts,
                communities: unsyncedCommunities,
                comments: unsyncedComments,
                users: unsyncedUsers,
                userSavedPosts: userSavedPosts.map((r: any) => ({
                    userId: r.user_id,
                    postId: r.post_id
                })),
                userCommunities: userCommunities.map((r: any) => ({
                    userId: r.user_id,
                    communityId: r.community_id
                })),
                userFollows: userFollows.map((r: any) => ({
                    followerId: r.follower_id,
                    followedId: r.followed_id
                }))
            };

            if (unsyncedPosts.length > 0 || unsyncedCommunities.length > 0 || unsyncedComments.length > 0 || unsyncedUsers.length > 0) {
                console.log("Enviando dados para o servidor...");
                const response = await api.post('/sync/push', pushPayload);
                console.log("Response do servidor:", response.status);

                if (unsyncedPosts.length > 0) {
                    await this.db.runAsync("UPDATE posts SET is_synced = 1 WHERE is_synced = 0");
                    console.log("Posts marcados como sincronizados");
                }
                if (unsyncedCommunities.length > 0) {
                    await this.db.runAsync("UPDATE communities SET is_synced = 1 WHERE is_synced = 0");
                    console.log("Comunidades marcadas como sincronizadas");
                }
                if (unsyncedComments.length > 0) {
                    await this.db.runAsync("UPDATE comments SET is_synced = 1 WHERE is_synced = 0");
                    console.log("Comentários marcados como sincronizados");
                }
                if (unsyncedUsers.length > 0) {
                    await this.db.runAsync("UPDATE users SET is_synced = 1 WHERE is_synced = 0");
                    console.log("Usuários marcados como sincronizados");
                }

                console.log("Push Sync finalizado com sucesso!");
            } else {
                console.log("Nenhum dado para sincronizar");
            }
        } catch (error) {
            console.error("Erro ao fazer Push Sync:", error);
            throw error;
        }
    }

    async pullSync() {
        try {
            console.log("Iniciando Pull Sync");
            
            const response = await api.get('/sync/pull');
            const data = response.data;

            console.log("Recebendo dados do servidor");

            for (const user of data.users || []) {
                try {
                    await this.db.runAsync(`
                        INSERT OR REPLACE INTO users (id, name, email, password, handle, bio, avatarUrl, bannerUrl, followers, following, is_synced)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
                    `, [
                        user.id,
                        user.name,
                        user.email,
                        user.password || '',
                        user.handle,
                        user.bio,
                        user.avatarUrl,
                        user.bannerUrl,
                        user.followers || 0,
                        user.following || 0
                    ]);
                } catch (e) {
                    console.warn("Erro ao inserir usuário:", user.id, e);
                }
            }
            if ((data.users || []).length > 0) {
                console.log(`${data.users.length} usuários sincronizados`);
            }

            for (const post of data.posts || []) {
                try {
                    await this.db.runAsync(`
                        INSERT OR REPLACE INTO posts (id, user_id, author, title, description, subject, tag, role, time, image_url, upvotes, comments_count, is_saved, created_at, updated_at, is_synced)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
                    `, [
                        post.id,
                        post.user_id,
                        post.author || "Autor",
                        post.title,
                        post.description,
                        post.subject,
                        post.tag,
                        post.role || "",
                        post.time || "agora",
                        post.image_url,
                        post.upvotes || 0,
                        post.comments_count || 0,
                        post.is_saved || 0,
                        post.created_at || new Date().toISOString(),
                        post.updated_at || new Date().toISOString()
                    ]);
                } catch (e) {
                    console.warn("Erro ao inserir post:", post.id, e);
                }
            }
            if ((data.posts || []).length > 0) {
                console.log(`${data.posts.length} posts sincronizados`);
            }

            for (const comm of data.communities || []) {
                try {
                    await this.db.runAsync(`
                        INSERT OR REPLACE INTO communities (id, name, members, description, is_joined, creator_id, banner_url, created_at, updated_at, is_synced)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
                    `, [
                        comm.id,
                        comm.name,
                        comm.members || "0",
                        comm.description,
                        comm.is_joined || 0,
                        comm.creator_id || 0,
                        comm.banner_url,
                        comm.created_at || new Date().toISOString(),
                        comm.updated_at || new Date().toISOString()
                    ]);
                } catch (e) {
                    console.warn("Erro ao inserir comunidade:", comm.id, e);
                }
            }
            if ((data.communities || []).length > 0) {
                console.log(`${data.communities.length} comunidades sincronizadas`);
            }

            for (const comment of data.comments || []) {
                try {
                    await this.db.runAsync(`
                        INSERT OR REPLACE INTO comments (id, post_id, user_id, author, avatar, text, time, likes, isLiked, replies, is_synced, created_at, updated_at)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?, ?)
                    `, [
                        comment.id,
                        comment.post_id,
                        comment.user_id || 0,
                        comment.author || "",
                        comment.avatar || "",
                        comment.text,
                        comment.time || "agora",
                        comment.likes || 0,
                        comment.isLiked || 0,
                        comment.replies || "[]",
                        comment.created_at || new Date().toISOString(),
                        comment.updated_at || new Date().toISOString()
                    ]);
                } catch (e) {
                    console.warn("Erro ao inserir comentário:", comment.id, e);
                }
            }
            if ((data.comments || []).length > 0) {
                console.log(`${data.comments.length} comentários sincronizados`);
            }

            for (const rel of data.userSavedPosts || []) {
                try {
                    await this.db.runAsync(
                        "INSERT OR IGNORE INTO user_saved_posts (user_id, post_id) VALUES (?, ?)",
                        [rel.userId, rel.postId]
                    );
                } catch (e) {
                    console.warn("Erro ao sincronizar saved post:", e);
                }
            }

            for (const rel of data.userCommunities || []) {
                try {
                    await this.db.runAsync(
                        "INSERT OR IGNORE INTO user_communities (user_id, community_id) VALUES (?, ?)",
                        [rel.userId, rel.communityId]
                    );
                } catch (e) {
                    console.warn("Erro ao sincronizar user community:", e);
                }
            }

            for (const rel of data.userFollows || []) {
                try {
                    await this.db.runAsync(
                        "INSERT OR IGNORE INTO user_follows (follower_id, followed_id) VALUES (?, ?)",
                        [rel.followerId, rel.followedId]
                    );
                } catch (e) {
                    console.warn("Erro ao sincronizar follow:", e);
                }
            }

            console.log("Pull Sync finalizado com sucesso!");
        } catch (error) {
            console.error("Erro ao fazer Pull Sync:", error);
            throw error;
        }
    }
}
